### DISCUSSION TITLE
Java Driver. https://github.com/ShaileshSurya/edgedb-java

### DISCUSSION THREAD

**Discussion by ShaileshSurya** [geldata/gel#1923] [Show and tell]
Hi,  I have created a primitive JAVA driver and published my changes here. https://github.com/ShaileshSurya/edgedb-java 

Were there efforts in this direction? Is it open for contribution?

**Comment by 1st1** [comment]
> Hi, I have created a primitive JAVA driver and published my changes here. https://github.com/ShaileshSurya/edgedb-java

This is amazing! 🎉

Thanks for working on it. 

> Were there efforts in this direction? Is it open for contribution?

No, we haven't yet started working on the Java driver. Please keep us in the loop and don't hesitate to ask questions or for help.  Our edgedb-python and edgedb-js drivers can be used as templates for the API design.

**Reply by ShaileshSurya** [reply]
Hi @1st1, 

Thank you for getting back. This is the first time I am writing a driver for a database, I am closely following up with the JavaScript and Python driver. As of now, the driver works with new connections, termination, and Simple Query Execution. 

I want to understand how the broader or high-level database driver works? Like how connection pulling, Error Tracking, Thread pooling works, and come into the picture? Do we have high-level documentation for the same? 

Also, I had a hard time understanding the Codecs in javascript and Python driver? Do we also have more documentation on this part? I skimmed through the EdgeDB documentation but could not figure out much.

**Reply by 1st1** [reply]
> Also, I had a hard time understanding the Codecs in javascript and Python driver? Do we also have more documentation on this part? I skimmed through the EdgeDB documentation but could not figure out much.

No, unfortunately we don't have high level documentation for that, except for https://www.edgedb.com/docs/internals/protocol/typedesc. Basically our types spec allows you to build a tree of coders/decoders - you build it as you parse the type spec. The edgedb-js code is relatively straightforward so I suggest to give it another look.

> I want to understand how the broader or high-level database driver works? Like how connection pulling, Error Tracking, Thread pooling works, and come into the picture? Do we have high-level documentation for the same?

I'd suggest to start with making the basic APIs work -- querying data and unmarshaling it into Java structs. And also to think how to make the basic protocol implementation compatible with blocking and non-blocking IO, although I really don't know what's the state of async ecosystem in Java.

**Reply by ShaileshSurya** [reply]
This answers. thank you.
